File: packages/android-kotlin/src/main/kotlin/com/onchainux/wallet/WalletManager.kt
/**
 * WalletManager — manages wallet connections and session lifecycle for Android.
 *
 * Handles wallet discovery, connection, disconnection, and state management.
 * Supports multiple connector types: injected, WalletConnect, Coinbase,
 * email, and social logins.
 *
 * ## Usage
 * ```kotlin
 * val manager = WalletManager()
 * manager.configure(config)
 *
 * val result = manager.connect("metamask")
 * println("Connected: ${result.account.address}")
 *
 * manager.disconnect()
 * ```
 */
package com.onchainux.wallet

import com.onchainux.core.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID

/**
 * Manages wallet connections and session state.
 */
class WalletManager {

    /** Currently connected account, if any. */
    private val _connectedAccount = MutableStateFlow<AccountInfo?>(null)
    val connectedAccount: StateFlow<AccountInfo?> = _connectedAccount.asStateFlow()

    /** Current connection status. */
    private val _connectionStatus = MutableStateFlow<ConnectionStatus>(ConnectionStatus.DISCONNECTED)
    val connectionStatus: StateFlow<ConnectionStatus> = _connectionStatus.asStateFlow()

    /** Available connectors. */
    private val _connectors = MutableStateFlow<List<ConnectorInfo>>(emptyList())
    val connectors: StateFlow<List<ConnectorInfo>> = _connectors.asStateFlow()

    /** Active session ID. */
    var sessionId: String? = null
        private set

    private var config: OnChainUXConfig? = null

    /**
     * Configure the wallet manager.
     */
    fun configure(config: OnChainUXConfig) {
        this.config = config
        _connectors.value = buildDefaultConnectors(config)
    }

    /**
     * Connect to a wallet by connector ID.
     *
     * @param connectorId The connector to use.
     * @return Connection result with account info.
     * @throws IllegalStateException if not configured.
     */
    suspend fun connect(connectorId: String): ConnectResult {
        _connectionStatus.value = ConnectionStatus.CONNECTING

        // Simulate wallet connection delay
        kotlinx.coroutines.delay(1000)

        val cfg = config ?: throw IllegalStateException("WalletManager not configured")

        val chainId = cfg.chains.firstOrNull()?.chainId ?: 1
        val symbol = cfg.chains.firstOrNull()?.nativeCurrency?.symbol ?: "ETH"

        val account = AccountInfo(
            address = "0x742d35Cc6634C0532925a3b844Bc9e7595f2bD18",
            balance = "1.234",
            chainId = chainId,
            chainSymbol = symbol
        )

        val newSessionId = UUID.randomUUID().toString()
        _connectedAccount.value = account
        sessionId = newSessionId
        _connectionStatus.value = ConnectionStatus.CONNECTED

        return ConnectResult(account = account, chainId = chainId, sessionId = newSessionId)
    }

    /**
     * Disconnect from the current wallet.
     */
    suspend fun disconnect() {
        _connectedAccount.value = null
        sessionId = null
        _connectionStatus.value = ConnectionStatus.DISCONNECTED
    }

    /**
     * Get the list of available connectors.
     */
    fun getConnectors(): List<ConnectorInfo> = _connectors.value

    /**
     * Switch to a different chain.
     */
    suspend fun switchChain(chainId: Int) {
        val cfg = config ?: throw IllegalStateException("WalletManager not configured")
        if (cfg.chains.none { it.chainId == chainId }) {
            throw IllegalArgumentException("Chain $chainId not supported")
        }

        _connectedAccount.value?.let { account ->
            val chain = cfg.chains.find { it.chainId == chainId }
            if (chain != null) {
                _connectedAccount.value = account.copy(
                    chainId = chainId,
                    chainSymbol = chain.nativeCurrency.symbol
                )
            }
        }
    }

    /**
     * Check if a wallet app is installed (via package manager).
     * In production, use PackageManager.resolveActivity() to check.
     */
    fun isWalletInstalled(walletId: String, context: android.content.Context): Boolean {
        val packageName = when (walletId) {
            "metamask" -> "io.metamask"
            "coinbase" -> "org.toshi"
            "trust" -> "com.wallet.crypto.trustapp"
            "rainbow" -> "me.rainbow"
            else -> null
        } ?: return false

        return try {
            context.packageManager.getPackageInfo(packageName, 0)
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun buildDefaultConnectors(config: OnChainUXConfig): List<ConnectorInfo> {
        return listOf(
            ConnectorInfo("metamask", "MetaMask", type = ConnectorType.WALLETCONNECT),
            ConnectorInfo("walletconnect", "WalletConnect", type = ConnectorType.WALLETCONNECT),
            ConnectorInfo("coinbase", "Coinbase Wallet", type = ConnectorType.COINBASE),
            ConnectorInfo("email", "Email", type = ConnectorType.EMAIL)
        )
    }
}

File: packages/ios-swift/Sources/OnChainUX/WalletManager.swift
/**
 * WalletManager — manages wallet connections and session lifecycle.
 *
 * Handles wallet discovery, connection, disconnection, and state management.
 * Supports multiple connector types: injected, WalletConnect, Coinbase,
 * email, and social logins.
 *
 * ## Usage
 * ```swift
 * let manager = WalletManager()
 * manager.configure(with: config)
 *
 * let result = try await manager.connect(connectorId: "metamask")
 * print("Connected: \(result.account.address)")
 *
 * await manager.disconnect()
 * ```
 */

import Foundation

/// Result of a successful wallet connection.
public struct ConnectResult: Sendable {
    /// Connected account info.
    public let account: AccountInfo
    /// Connected chain ID.
    public let chainId: Int
    /// Session ID.
    public let sessionId: String
    
    public init(account: AccountInfo, chainId: Int, sessionId: String) {
        self.account = account
        self.chainId = chainId
        self.sessionId = sessionId
    }
}

/// Manages wallet connections and session state.
public final class WalletManager: ObservableObject {
    
    /// Currently connected account, if any.
    @Published public private(set) var connectedAccount: AccountInfo?
    
    /// Current connection status.
    @Published public private(set) var connectionStatus: ConnectionStatus = .disconnected
    
    /// Available connectors.
    @Published public private(set) var connectors: [ConnectorInfo] = []
    
    /// Active session ID.
    public private(set) var sessionId: String?
    
    private var config: OnChainUXConfig?
    
    // MARK: - Lifecycle
    
    /// Configure the wallet manager.
    /// - Parameter config: App configuration.
    public func configure(with config: OnChainUXConfig) {
        self.config = config
        connectors = buildDefaultConnectors(config: config)
    }
    
    /// Connect to a wallet by connector ID.
    /// - Parameter connectorId: The connector to use.
    /// - Returns: Connection result with account info.
    /// - Throws: OnChainUXError if connection fails.
    public func connect(connectorId: String) async throws -> ConnectResult {
        connectionStatus = .connecting
        
        // Simulate wallet connection (in production, integrate with WalletConnect SDK or injected providers)
        try await Task.sleep(nanoseconds: 1_000_000_000)
        
        guard let config = config else {
            throw OnChainUXError.notConfigured
        }
        
        let chainId = config.chains.first?.chainId ?? 1
        let symbol = config.chains.first?.nativeCurrency.symbol ?? "ETH"
        
        // Simulated connection result
        let account = AccountInfo(
            address: "0x742d35Cc6634C0532925a3b844Bc9e7595f2bD18",
            balance: "1.234",
            chainId: chainId,
            chainSymbol: symbol
        )
        
        let newSessionId = UUID().uuidString
        connectedAccount = account
        sessionId = newSessionId
        connectionStatus = .connected
        
        return ConnectResult(account: account, chainId: chainId, sessionId: newSessionId)
    }
    
    /// Disconnect from the current wallet.
    public func disconnect() async {
        connectedAccount = nil
        sessionId = nil
        connectionStatus = .disconnected
    }
    
    /// Get the list of available connectors.
    /// - Returns: Array of connector info.
    public func getConnectors() -> [ConnectorInfo] {
        connectors
    }
    
    /// Check if a specific wallet app is installed (via URL scheme).
    /// - Parameter walletId: Wallet identifier.
    /// - Returns: Whether the wallet app appears to be installed.
    public func isWalletInstalled(walletId: String) -> Bool {
        guard let scheme = urlScheme(for: walletId) else { return false }
        guard let url = URL(string: scheme) else { return false }
        return UIApplication.shared.canOpenURL(url)
    }
    
    /// Switch to a different chain.
    /// - Parameter chainId: Target chain ID.
    public func switchChain(chainId: Int) async throws {
        guard let config = config else {
            throw OnChainUXError.notConfigured
        }
        guard config.chains.contains(where: { $0.chainId == chainId }) else {
            throw OnChainUXError.chainNotSupported(chainId)
        }
        
        // In production: send wallet_switchEthereumChain request to connected wallet
        if var account = connectedAccount {
            // Update chain symbol
            if let chain = config.chains.first(where: { $0.chainId == chainId }) {
                account = AccountInfo(
                    address: account.address,
                    balance: account.balance,
                    chainId: chainId,
                    chainSymbol: chain.nativeCurrency.symbol,
                    ensName: account.ensName
                )
                connectedAccount = account
            }
        }
    }
    
    // MARK: - Private
    
    private func buildDefaultConnectors(config: OnChainUXConfig) -> [ConnectorInfo] {
        var list: [ConnectorInfo] = [
            ConnectorInfo(id: "metamask", name: "MetaMask", type: .walletconnect),
            ConnectorInfo(id: "walletconnect", name: "WalletConnect", type: .walletconnect),
            ConnectorInfo(id: "coinbase", name: "Coinbase Wallet", type: .coinbase),
            ConnectorInfo(id: "email", name: "Email", type: .email),
        ]
        return list
    }
    
    private func urlScheme(for walletId: String) -> String? {
        switch walletId {
        case "metamask": return "metamask://"
        case "coinbase", "cbwallet": return "cbwallet://"
        case "rainbow": return "rainbow://"
        case "trust": return "trust://"
        case "rabby": return "rabby://"
        case "phantom": return "phantom://"
        default: return nil
        }
    }
}
